tests/distributions/test_hmm.py

```diff
@@ -164,3 +185,81 @@ def model(data):
+    ((11,), (11, 7), (11, 7)),
+    ((4, 1, 1), (3, 1, 7), (2, 7)),
+], ids=str)
+def test_gaussian_mrf_shape(init_shape, trans_shape, obs_shape, hidden_dim, obs_dim):
+    init_dist = random_mvn(init_shape, hidden_dim)
+    trans_dist = random_mvn(trans_shape, hidden_dim + hidden_dim)
+    obs_dist = random_mvn(obs_shape, hidden_dim + obs_dim)
+    d = dist.GaussianMRF(init_dist, trans_dist, obs_dist)
+
+    shape = broadcast_shape(init_shape + (1,), trans_shape, obs_shape)
+    expected_batch_shape, time_shape = shape[:-1], shape[-1:]
+    expected_event_shape = time_shape + (obs_dim,)
+    assert d.batch_shape == expected_batch_shape
+    assert d.event_shape == expected_event_shape
+
+    data = obs_dist.expand(shape).sample()[..., hidden_dim:]
+    assert data.shape == d.shape()
+    actual = d.log_prob(data)
+    assert actual.shape == expected_batch_shape
+
+
+@pytest.mark.parametrize('sample_shape', [(), (5,)], ids=str)
+@pytest.mark.parametrize('batch_shape', [(), (4,), (3, 2)], ids=str)
+@pytest.mark.parametrize('obs_dim', [1, 2])
+@pytest.mark.parametrize('hidden_dim', [1, 2])
+@pytest.mark.parametrize('num_steps', [1, 2, 3, 4])
+def test_gaussian_mrf_log_prob(sample_shape, batch_shape, num_steps, hidden_dim, obs_dim):
+    init_dist = random_mvn(batch_shape, hidden_dim)
+    trans_dist = random_mvn(batch_shape + (num_steps,), hidden_dim + hidden_dim)
+    obs_dist = random_mvn(batch_shape + (num_steps,), hidden_dim + obs_dim)
+    d = dist.GaussianMRF(init_dist, trans_dist, obs_dist)
+    data = obs_dist.sample(sample_shape)[..., hidden_dim:]
+    assert data.shape == sample_shape + d.shape()
+    actual_log_prob = d.log_prob(data)
+
+    # Compare against hand-computed density.
+    # We will construct enormous unrolled joint gaussians.
+    T = num_steps
+    init = mvn_to_gaussian(init_dist)
+    trans = mvn_to_gaussian(trans_dist)
+    obs = mvn_to_gaussian(obs_dist)
+
+    unrolled_trans = reduce(operator.add, [
+        trans[..., t].event_pad(left=t * hidden_dim, right=(T - t - 1) * hidden_dim)
+        for t in range(T)
+    ])
+    unrolled_obs = reduce(operator.add, [
+        obs[..., t].event_pad(left=t * obs.dim(), right=(T - t - 1) * obs.dim())
+        for t in range(T)
+    ])
+    # Permute from HOHOHO to HHHOOO.
+    perm = torch.cat([torch.arange(hidden_dim) + t * obs.dim() for t in range(T)] +
+                     [torch.arange(obs_dim) + hidden_dim + t * obs.dim() for t in range(T)])
+    unrolled_obs = unrolled_obs.event_permute(perm)
+    unrolled_data = data.reshape(data.shape[:-2] + (T * obs_dim,))
+
+    assert init.dim() == hidden_dim
+    assert unrolled_trans.dim() == (1 + T) * hidden_dim
+    assert unrolled_obs.dim() == T * (hidden_dim + obs_dim)
+    joint = gaussian_tensordot(init, unrolled_trans, hidden_dim)
+    joint = gaussian_tensordot(joint, unrolled_obs, T * hidden_dim)
+    expected_log_prob = joint.log_density(unrolled_data)
+    assert_close(actual_log_prob, expected_log_prob)
```

Comment:
This test assumes that we trust the math of tensordot. How about we add one more test for the case `obs_dist` is a diagonal block MVN? That implies obs is independent of hidden, so I think that in that case, `expected_log_prob = obs_dist.log_prob(value).sum(-1)`. This helps us detect if there are missing constants.

Reply:
I've added a test like that, but it is a little more complex due to the MRF factors combining to a non-normalized distribution (which wouldn't have been needed if this had been an HMM).

Reply:
On second thought, I guess we'd like the `GaussianGaussianMRF` distribution to be normalized (it is not currently normalized). I think we could do this by performing two contractions: the current contraction, and another contraction marginalizing over observations, i.e. `p(obs) = p(obs, hidden) / p(hidden)` WDYT?

Reply:
I think we'll want to normalize. Here's a potential implementation (I'll work on tests):
```py
def log_prob(self, value):
    # We compute a normalized distribution as p(obs,hidden) / p(hidden).
    # Combine observation and transition factors.
    logp_h = self._trans
    logp_oh = self._trans
    logp_oh += self._obs.condition(value).event_pad(left=self.hidden_dim)

    # Concatenate p(hidden) and p(observed,hidden) into a single Gaussian.
    batch_shape = logp_oh.batch_shape
    logp = Gaussian.cat([logp_h.expand(batch_shape),
                         logp_oh.expand(batch_shape)])

    # Eliminate time dimension.
    logp = _sequential_gaussian_tensordot(logp)

    # Combine initial factor.
    logp = gaussian_tensordot(self._init, logp, dims=self.hidden_dim)

    # Marginalize out final state.
    logp = logp.event_logsumexp()
    logp_h, logp_oh = logp[0], logp[1]
    return logp_oh - logp_h  # = log( p(obs,hidden) / p(hidden) )
```

Reply:
> p(obs) = p(obs, hidden) / p(hidden)

Agree, this would account for normalization.